lib/Dialect/StandardOps/Ops.cpp

```diff
@@ -2018,6 +2018,61 @@ OpFoldResult SelectOp::fold(ArrayRef<Attribute> operands) {
   return nullptr;
 }
 
+//===----------------------------------------------------------------------===//
+// SplatOp
+//===----------------------------------------------------------------------===//
+
+static void print(OpAsmPrinter &p, SplatOp op) {
+  p << "splat " << *op.getOperand();
+  p.printOptionalAttrDict(op.getAttrs());
+  p << " : " << op.getType();
+}
+
+static ParseResult parseSplatOp(OpAsmParser &parser, OperationState &result) {
+  OpAsmParser::OperandType splatValueInfo;
+  ShapedType shapedType;
+
+  return failure(parser.parseOperand(splatValueInfo) ||
+                 parser.parseOptionalAttributeDict(result.attributes) ||
+                 parser.parseColonType(shapedType) ||
+                 parser.resolveOperand(splatValueInfo,
+                                        shapedType.getElementType(),
+                                        result.operands) ||
+                 parser.addTypeToList(shapedType, result.types));
+}
+
+static LogicalResult verify(SplatOp op) {
+  if (!op.getOperand()->getType().isIntOrIndexOrFloat())
+    return op.emitError("operand should be of integer, index, or float type");
+
+  auto type = op.getType();
+
+  if (!type.isa<VectorType>() && !type.isa<TensorType>())
+    return op.emitError("result type should either be a vector or tensor type");
+
+  if (op.getOperand()->getType() != type.cast<ShapedType>().getElementType())
+    return op.emitError("operand should be of elemental type of result type");
+
+  return success();
+}
+
+// Constant folding hook for SplatOp.
+OpFoldResult SplatOp::fold(ArrayRef<Attribute> operands) {
+  assert(operands.size() == 1 && "splat takes one operand");
+
+  auto constOperand = operands.front();
+  if (!constOperand.dyn_cast_or_null<IntegerAttr>() &&
+      !constOperand.dyn_cast_or_null<FloatAttr>())
+    return {};
+
+  auto shapedType = getType().cast<ShapedType>();
+  assert(shapedType.getElementType() == constOperand.getType() &&
+         "incorrect input attribute type for folding");
+
+  // SplatElementsAttr::get treats single value for second arg as being a splat.
+  return SplatElementsAttr::get(shapedType, {constOperand});
```

Comment:
You can't have a splat of index type.

Reply:
You can actually have a splat of index type. Addresses can be stored in a vector and lowered.

Reply:
What I'm saying is, you can't create a SplatElementsAttr(technically a DenseElementsAttr) with an attribute of IndexType. It will assert.

Reply:
Thanks. We use 64-bit APInt to store attributes for the index type:
https://github.com/tensorflow/mlir/blob/c49b963db45a8425371083dc3ddef860bd28ad5c/lib/IR/Attributes.cpp#L268
It should be okay to add the missing case in DenseElementsAttr::get?

Reply:
We could, but if we go down this route we should really clarify how we store constants of IndexType internally. It implies that an 'index' has a maximum representation of i64, and must be treated as i64 when manipulating via constants. If someone writes code with their target in mind(i32 for example), index type won't work as they expect. DenseElementsAttr suffers more from this because we highly discourage operating on Attributes, because they are expensive to construct, and instead operate directly on APInt/uint32_t/uint64_t/etc.

Reply:
This is part of the reason why you are *not* allowed to use index in a vector or tensor. As such, index type should be removed from this op.
https://github.com/tensorflow/mlir/blob/master/g3doc/LangRef.md#index-type

Reply:
> We could, but if we go down this route we should really clarify how we store constants of IndexType internally. It implies that an 'index' has a maximum representation of i64, and must 

We already chose to store index type constants using 64-bit APInt internally. This was discussed at some point, and I think the thought process was that if a target didn't support 64-bit addresses, we would just truncate during target codegen. There were two other alternative design points that were discussed but this was the one that we went with at that time. The discussion on the choices and trade-offs was from @lattner

Reply:
> This is part of the reason why you are _not_ allowed to use index in a vector or tensor. As such, index type should be removed from this op.
> https://github.com/tensorflow/mlir/blob/master/g3doc/LangRef.md#index-type

Thanks - I had forgotten about this. Then, we don't need to support index type here at all.

Reply:
Done.